### DISCUSSION TITLE
About import in starter

### DISCUSSION THREAD

**Discussion by madjxatw** [LazyVim/LazyVim#339] [General]
I find `{ "LazyVim/LazyVim", import = "lazyvim.plugins" }` in https://github.com/LazyVim/starter/blob/main/lua/config/lazy.lua, and feel a bit confused about how it exactly works.

As per what stated in https://github.com/folke/lazy.nvim#readme, lazy.nvim by default should call `require("LazyVim").setup(opts)` when LazyVim gets loaded, but doesn't `LazyVim/init.lua` just simply quit vim?  Where does the `setup()` come from?  I find a `setup()` function defined in `LazyVim/lua/lazyvim/init.lua` which actually invokes the `setup()` functions from `LazyVim/lua/lazyvim/config/init.lua` to load keymaps and autocmds, it is supposed to be called but I could find nowhere it is called. 

My understanding is that `import = "lazyvim.plugins"` sources all lua files under `LazyVim/lua/lazyvim/plugins/` but I find only `core.lua` calls `require("lazyvim.config").init()` which loads options. What is the magic behind `{ "LazyVim/LazyVim", import = "lazyvim.plugins" }` that gets `setup()` from `LazyVim/lua/lazyvim/init.lua` called?

**Comment by dpetka2001** [comment]
In short `Lazy` looks for plugins to load in your own `lua/plugins` directory. If you want to load something that is in a different path you should `import` it. That's why you import `lazyvim.plugins`, because it's in another directory. That's how I understand it at least, you could wait for a more discriptive answer from the maintainer after he returns from his vacation.

**Reply by madjxatw** [reply]
I think there is some trick inside lazy.vim that takes `lazyvim` as a plugin and call its `setup()`, but have no time to figure out further.

**Comment by folke** [comment]
https://github.com/LazyVim/LazyVim/blob/befa6c67a4387b0db4f8421d463f5d03f91dc829/lua/lazyvim/plugins/core.lua#L5

That `config=true` will call `setup`.

`import` is unrelated to setup and will just load some plugin specs on the rtp, in this case the ones from LazyVim

**Reply by madjxatw** [reply]
`LazyVim/init.lua` has no `setup()` defined but `LazyVim/lua/layzvim/init.lua` does. How does the `setup()` from `LazyVim/lua/lazyvim/init.lua` get populated to `LazyVim`? What actually gets `lazyvim` involved in `
 { "LazyVim/LazyVim", priority = 10000, lazy = false, config = true, version = "*" }`.

**Reply by folke** [reply]
Config=true. Check the lazy.nvim docs. This is the same with any plugin. There's no special handling for lazyvim

**Reply by madjxatw** [reply]
I've read lazy.nvim docs but still didn't get where the `setup()` comes from.

>The default implementation will automatically run require("plugin").setup(opts).

Given ` { "LazyVim/LazyVim", priority = 10000, lazy = false, config = true, version = "*" }`, lazy.nvim should call `require("LazyVim").setup(opts)` but not `require("LazyVim.lazyvim").setup()`.  There is no `setup()` defined in `LazyVim/init.lua`.

**Reply by mdmjmsa** [reply]
@madjxatw, I think you might a little bit misunderstand how a lua plugin works. As my understanding, Neovim will load each `lua` file and subdirectories containing a `init.lua` file under the `lua` directory of a plugin, the same applies to how `setup()` for LazyVim is found in `LazyVim/lua/lazyvim/init.lua`, that is `require("LazyVim")` loads `LazyVim/lua/lazyvim/init.lua` and finds `setup()` there.

**Reply by loiccoyle** [reply]
The setup function is defined here: https://github.com/LazyVim/LazyVim/blob/592074ad802be2462306d3991024f350571dfab2/lua/lazyvim/init.lua#L3-L6

As far as I know, calling `require("LazyVim")` loads the module which starts at https://github.com/LazyVim/LazyVim/tree/main/lua/lazyvim so `require("LazyVim").setup()` calls the function defined above.

If this is still unclear you should read up on how neovim lua plugins work.

**Reply by madjxatw** [reply]
Thank you all for helping me finally get how a neovim lua plugins works, really appreciated! BTW, LazyVim is awesome.
